### PR TITLE
Simplify tilegrid API and internals

### PR DESCRIPTION
**Pull request by ahocevar**
From talking clients through all iterations of the tilegrid API when using custom tile url functions, and from experiences with the ol3-cesium integration of tile layers, it became clear that things are still not straightforward:

* Tile coordinates for WMS and ArcGISRest increase from bottom to top, but those used by XYZ tile grids increase from top to bottom. This requires [different code paths when using the tile url function](https://github.com/ahocevar/ol3-cesium/commit/5d30725128a1c9b295883b3f23cbe5ceb8b3d765#diff-40e38a98ab85558fe800cc965e85730fR210). The fix from this pull request is to remove tile coord transforms altogether, so application developers always know that `ol.TileCoord` has y values increasing from bottom to top.
* XYZ tile grids use a top-left origin, all other tile grids use a bottom-left origin. With tile coord transforms removed, this means that custom tile url functions get different tile coordinates for different sources and tile grids. This pull request changes all tile grids that OpenLayers creates to use the top-left corner of the extent as origin.

With these API simplifications, it is also possible to simplify tile grid handling internally. The Zoomify tile grid can finally be removed, because `ol.tilegrid.TileGrid` now provides everything that the Zoomify tile source needs.

### CODE REVIEW

**Comment by bjornharrtell** [comment]
Nice!

File: changelog/upgrade-notes.md [hunk 1]
```diff
@@ -36,13 +36,22 @@ Note that `ol.FeatureOverlay#getFeatures()` returned an `{ol.Collection.<ol.Feat
...
+Until now, the API exposed two different types of `ol.TileCoord` tile coordinates: internal ones that increase from left to right and from bottom to top, and transformed ones that may increase from top to bottom, as defined by a transform function on the tile grid. With this change, the API now only exposes tile coordinates that increase from left to right and from bottom to top.
 
-The first `tileCoord` argument of `ol.TileUrlFunctionType` now expects transformed tile coordinates instead of internal OpenLayers tile coordinates. Accordingly, `ol.tilegrid.TileGrid#getTileCoordForCoordAndZ` and `ol.tilegrid.TileGrid#getTileCoordForCoordAndResolution` now return transformed tile coordinates.
+Previously, tile grids created by OpenLayers either had their origin at the top-left or at the bottom-left corner of the tile grid. To make it easier to for application developers to transform tile coordinates to the common XYZ tiling scheme, all tile grids that OpenLayers creates internally have their origin now at the top-left corner.
```
Comment (elemoine) [inline changelog/upgrade-notes.md]: Don't you mean `bottom-left` here?

Comment (ahocevar) [inline changelog/upgrade-notes.md]: No, it's the top-left corner of the extent now. But the tile coordinates increase upwards.

Comment (elemoine) [inline changelog/upgrade-notes.md]: "of the extent"  was what I was missing. I got it.

File: src/ol/tilegrid/tilegrid.js [hunk 2]
```diff
@@ -427,9 +391,10 @@ ol.tilegrid.TileGrid.prototype.getTileCoordForXYAndResolution_ = function(
   var origin = this.getOrigin(z);
   var tileSize = ol.size.toSize(this.getTileSize(z), this.tmpSize_);
 
-  var adjust = reverseIntersectionPolicy ? 0.5 : 0;
-  var xFromOrigin = ((x - origin[0]) / resolution + adjust) | 0;
-  var yFromOrigin = ((y - origin[1]) / resolution + adjust) | 0;
+  var adjustX = reverseIntersectionPolicy ? 0.5 : 0;
+  var adjustY = reverseIntersectionPolicy ? 0 : 0.5;
+  var xFromOrigin = Math.floor((x - origin[0]) / resolution + adjustX);
+  var yFromOrigin = Math.floor((y - origin[1]) / resolution + adjustY);
```
Comment (elemoine) [inline src/ol/tilegrid/tilegrid.js]: Why did you change this?

**Comment by elemoine** [comment]
I just added questions for me to understand. I think this no-surprises branch is just great. Please merge.

File: changelog/upgrade-notes.md (hunk 1 above)
Comment (probins) [inline changelog/upgrade-notes.md]: /easier to for/easier for/

File: src/ol/tilegrid/tilegrid.js (hunk 2 above)
Comment (ahocevar) [inline src/ol/tilegrid/tilegrid.js]: Two reasons:

1. The change from `(whatever) | 0` to `Math.floor(whatever)` was necessary because `(Infinity) | 0` is `0`, but should be `Infinity`.
1. The `y`adjustment for adjusting to screen pixels was previously incorrect, because screen coordinates increase downwards, but map coordinates increase upwards. The tests caught this problem after changing the origin to the top-left corner of the extent.

**Comment by ahocevar** [comment]
Thanks for the review @elemoine. I'll add 'of the extent' to the upgrade notes, so it will be easier for others to understand the change too. I'll also change the wording 'from bottom to top' to 'upwards' and 'from top to bottom' to downwards, as @tschaub suggested yesterday. Then I'll merge.

File: changelog/upgrade-notes.md [hunk 3]
```diff
@@ -36,13 +36,22 @@ Note that `ol.FeatureOverlay#getFeatures()` returned an `{ol.Collection.<ol.Feat
...
+Until now, the API exposed two different types of `ol.TileCoord` tile coordinates: internal ones that increase left to right and upward, and transformed ones that may increase downward, as defined by a transform function on the tile grid. With this change, the API now only exposes tile coordinates that increase left to right and upward.
 
-The first `tileCoord` argument of `ol.TileUrlFunctionType` now expects transformed tile coordinates instead of internal OpenLayers tile coordinates. Accordingly, `ol.tilegrid.TileGrid#getTileCoordForCoordAndZ` and `ol.tilegrid.TileGrid#getTileCoordForCoordAndResolution` now return transformed tile coordinates.
+Previously, tile grids created by OpenLayers either had their origin at the top-left or at the bottom-left corner of the extent. To make it easier to for application developers to transform tile coordinates to the common XYZ tiling scheme, all tile grids that OpenLayers creates internally have their origin now at the top-left corner of the extent.
```
Comment (probins) [inline changelog/upgrade-notes.md]: /easier to for/easier for/
